endpoint: refactor Docker + IPVLAN interactions

ianvernon opened the pull request:
The endpoint's internals were leaked in the `finishIpvlanInit` function. Refactor this function to be part of the endpoint, and hide the implementation of the actual IPVLAN + netNS interactions behind a callback.

A consequence of this refactoring is that certain functions no longer need to be exported in both `pkg/endpoint/connector` and `pkg/endpoint`. 

Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8974)
<!-- Reviewable:end -->

Commit:
endpoint: refactor Docker + IPVLAN interactions

The endpoint's internals were leaked in the `finishIpvlanInit`
function. Refactor this function to be part of the endpoint, and hide the
implementation of the actual IPVLAN + netNS interactions behind an interface as
to not leak implementation of the datapath itself into `pkg/endpoint`.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
connector: do not export SetupIpvlanInRemoteNs

This function isn't used outside of `pkg/endpoint/connector` any more.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: do not export IPVLAN configuration functions

This functionality is now contained within `pkg/endpoint`, so no need to export
it anymore.

Signed-off by: Ian Vernon <ian@cilium.io>

raybejjani commented on pkg/endpoint/bpf.go:
```diff
@@ -1146,3 +1147,62 @@ func (e *Endpoint) RequireRouting() (required bool) {
 	return e.DatapathConfiguration.InstallEndpointRoute
 }
+
+type netNSManager interface {
```
Should this just be part of datapath.Datapath? (alternatively, some sort of DatapathOps? I would expect to capture more things into it someday, like the `bpf.LocalMapPath` wrappers at the start of this function).

raybejjani replied:
Oh, the `d` here is not the daemon... and piping a `Datapath` to here won't work. sigh... I don't suppose I could convince you to make this a callback instead of of adding all the boilerplate?

ianvernon replied:
What do you envision for the callback exactly? I'm unsure where I would plumb it into `pkg/workloads`. Can you provide an example of what you want?

brb replied:
Can't we just call `connector.ConfigureNetNSForIPVLAN()` without the indirection via the interface/struct from `(*Endpoint) FinishIPVLANInit()`?

raybejjani replied:
I just meant instead of passing around `netNSManager` objects pass about `func (netNsPath string) (mapFD, mapID int, err error)` directly. This is definitely a style thing. I just don't like types that just exist to pass functions about.

ianvernon replied:
Got it, makes sense!

ianvernon replied:
Done!

raybejjani commented on pkg/endpoint/connector/docker.go:
```diff
@@ -0,0 +1,78 @@
+// ConfigureNetNSForIPVLAN sets up IPVLAN in the specified network namespace.
+// Returns the file descriptor for the tail call map / ID, and an error if
+// any operation while configuring said namespace fails.
+func (d *DockerNetNSConfigurer) ConfigureNetNSForIPVLAN(netNsPath string) (mapFD, mapID int, err error) {
```
We already have a `pkg/endpoint/ipvlan.go` that seems to have same-ish enough functions for this to go in there. I'm not sure how docker-specific this function is and if you wanted to highlight that part more clearly (perhaps you have more docker-specific things in mind for later?)

ianvernon replied:
Agreed, function will be moved to `ipvlan.go`.

raybejjani commented on pkg/endpoint/bpf.go:
```diff
@@ -1146,3 +1147,62 @@ func (e *Endpoint) RequireRouting() (required bool) {
+
+	// No need to finish IPVLAN initialization for Docker if the endpoint isn't
+	// running with Docker.
+	if e.GetDockerNetworkID() == "" {
```
Since we lock below, it would be more correct to do the check while holding the same lock. The current implementation races in the PR and master.
I realise that the current code has the same race, however so perhaps leave a note or something.

ianvernon replied:
`GetDockerNetworkID` acquires the Endpoint's mutex, but I agree that having one critical section per function is nice. Will access the field directly while holding the lock in the next upload.

raybejjani commented on pkg/workloads/docker.go:
```diff
@@ -563,89 +558,3 @@ func (d *dockerClient) GetAllInfraContainersPID() (map[string]int, error) {
-func finishIpvlanInit(ep *endpoint.Endpoint, netNsPath string) error {
-	var ipvlanIface string
-
-	if netNsPath == "" {
```
We lost this nil-check. Are we now guaranteed that it's non-nil?

ianvernon replied:
Good catch! Will be fixed in next upload.

raybejjani reviewed (changes requested):
This is mergeable, but I'm requesting changes so it doesn't get yolo-merged before you can decide on whether to fix the race around GetDockerNetworkID. Feel free to merge if you wanted to do it in another PR.

Commit:
ipvlan: pass callback to configure namespace instead of type

The wrapper type had no fields, so just pass the function itself instead.

Signed-off by: Ian Vernon <ian@cilium.io>

brb commented on pkg/endpoint/bpf.go:
```diff
@@ -1146,3 +1147,63 @@ func (e *Endpoint) RequireRouting() (required bool) {
+		return nil
+	}
+
+	mapFD, mapID, err := netNSIPVLANConfigFunc(netNsPath)
```
Can't we just call `connector.ConfigureNetNSForIPVLAN()`? Then we wouldn't need to pass the callback to this method.

ianvernon replied:
I want to avoid passing in netlink as a transitive dependency into the endpoint package. It makes running unit tests require Linux. My overarching goal with the refactors I have been performing is to twofold:

* to reduce exposure of Endpoint internals to outside packages to allow for easier testing of its lifecycle and refactoring of its internals without having to worry about the impact on callers
* to reduce the dependency on explicit datapath implementations which would make testing of the Endpoint more difficult

brb replied:
> It makes running unit tests require Linux.

Probably we could create `pkg/endpoint/connector/ipvlan_{linux,darwin}.go` or add the build constraint `// +build linux`` in a unit test module.

The problem I see with passing the callback is that we leak the internal implementation detail to a user of the function. Also, it gives a wrong impression to the user that there exists multiple different ways to configure netns and friends for the ipvlan.

ianvernon replied:
The nice thing about providing an interface or a callback vs. having files with architectures that determine which version of a file (and thus its dependencies) that are pulled into the resultant binary is that I could provide a function to `FinishIPVLANInit` in a unit test that implements the `netNSIPVLANConfigFunc` type in the following way:

```
func errorConfigFunc(netNsPath string) (mapFD, mapID int, err error) {
   return 0, 0, fmt.Errorf("error case")
}
```

Now I have a way of injecting a failure case into our testing; as far as I know, I will not be able to do this with your suggestion.

There is only one way of configuring a netns for IPVLAN (via netlink), but we have no way of testing its failure conditions.

What are your thoughts on me adding this functionality to the `pkg/datapath:Datapath` interface?

brb replied:
I see your intention. As long as we have such unit test with failure conditions, the indirection makes sense. However, it comes with a price that for a reader it's more difficult to follow the ipvlan setup which is already quite complicated.

> What are your thoughts on me adding this functionality to the pkg/datapath:Datapath interface?

If the Datapath interface can accommodate a custom ipvlan-related method and an instance of the interface can be accessed from the Endpoint, then yes. Or maybe it's time `type Connector interface`?

ianvernon replied:
@brb check the latest commit - I believe it addresses your concerns as well as puts it behind an interface.

brb replied:
Amazing!

Commit:
hide IPVLAN Setup behind interface

Signed-off by: Ian Vernon <ian@cilium.io>
